@kriscendobot CI lint failure:

https://github.com/endojs/endo-but-for-bots/actions/runs/27721968884/job/82009426787?pr=452


In evoke/config.sh line 1:

^-- SC2148 (error): Tips depend on target shell and yours is unknown. Add a shebang or a 'shell' directive.


In evoke/config.sh line 8:
NOTIFY=http://127.0.0.1:8077/chat
^----^ SC2034 (warning): NOTIFY appears unused. Verify use (or export if used externally).


In evoke/config.sh line 16:
AGENT_NAME='claude'
^--------^ SC2034 (warning): AGENT_NAME appears unused. Verify use (or export if used externally).


In evoke/config.sh line 17:
AGENT_ARGS=('claude' '--dangerously-skip-permissions')
^--------^ SC2034 (warning): AGENT_ARGS appears unused. Verify use (or export if used externally).


In evoke/config.sh line 18:
AGENT_SESSIONS="$HOME/.claude/projects/${PWD//\//-}"
^------------^ SC2034 (warning): AGENT_SESSIONS appears unused. Verify use (or export if used externally).


In evoke/config.sh line 19:
AGENT_CURRENT_SESSION=
^-------------------^ SC2034 (warning): AGENT_CURRENT_SESSION appears unused. Verify use (or export if used externally).


In packages/fae/provider-setup.sh line 15:
set -a; source "$ENV_FILE"; set +a
               ^---------^ SC1090 (warning): ShellCheck can't follow non-constant source. Use a directive to specify location.


In packages/genie/test/integration.sh line 49:
      source "$1"
             ^--^ SC1090 (warning): ShellCheck can't follow non-constant source. Use a directive to specify location.


In packages/genie/test/integration.sh line 60:
      export "$1"
             ^--^ SC2163 (warning): This does not export '1'. Remove $/${} for that, or use ${var?} to quiet.


In packages/genie/test/integration.sh line 327:
for i in $(seq 1 30); do
^-^ SC2034 (warning): i appears unused. Verify use (or export if used externally).


In packages/jaine/provider-setup.sh line 23:
set -a; source "$ENV_FILE"; set +a
               ^---------^ SC1090 (warning): ShellCheck can't follow non-constant source. Use a directive to specify location.


In packages/jaine/reload.sh line 49:
  set -a; source "$ENV_FILE"; set +a
                 ^---------^ SC1090 (warning): ShellCheck can't follow non-constant source. Use a directive to specify location.

For more information:
  https://www.shellcheck.net/wiki/SC2148 -- Tips depend on target shell and y...
  https://www.shellcheck.net/wiki/SC1090 -- ShellCheck can't follow non-const...
  https://www.shellcheck.net/wiki/SC2034 -- AGENT_ARGS appears unused. Verify...
Error: Process completed with exit code 123.

--- a/packages/daemon/src/daemon.js
+++ b/packages/daemon/src/daemon.js
@@ -5139,7 +5139,15 @@ const makeDaemonCore = async (
           console.log(
             `Endo daemon peer node ${nodeId.slice(0, 8)} connection disposed`,
           );
-          dropLiveValue(context.id);
+          // Cancel the peer formula's context so the formula-lifecycle
+          // machinery tears down and drops all dependent remote presences
+          // via thisDiesIfThatDies.  The next use of any remote presence
+          // reincarnates the peer formula and re-dials from scratch.
+          // dropLiveValue alone was insufficient: it removed the peer from
+          // the live-value cache but left the formula context alive, so
+          // dependent remote presences were not revoked and the stale
+          // currentGatewayP prevented re-dial on subsequent use.
+          context.cancel(new Error('peer connection lost'));
         },
       );
     };
@@ -5219,6 +5227,20 @@ const makeDaemonCore = async (
     // If crossed-hellos accept bias abandons our dial, retry once.
     // The state machine should now be in `accepted` state, and the
     // retry will return the already-accepted gateway without dialing.
+    //
+    // TODO(option-a-simplification): After the dispose-callback change
+    // to context.cancel (Option A), the peer formula's context is
+    // cancelled on any connection loss.  That means this formula
+    // instance is torn down before it could ever see a second
+    // connection.  The `isAbandonError` predicate and the resilient-dial
+    // retry inside `resilientDial` are therefore unreachable for any
+    // post-connect abandon.  They remain for the initial-dial crossed-
+    // hellos case (where the state machine rejects our outgoing dial
+    // while accepting an inbound one, producing an abandon error before
+    // the first successful connection).  A follow-up can simplify by
+    // removing the `isAbandonError` catch in `ResilientPeerGateway.
+    // provide` and collapsing `currentGatewayP` to a plain `dialAttempt`
+    // call.
     const isAbandonError = err =>
       err &&
       typeof err.message === 'string' &&
@@ -5243,12 +5265,14 @@ const makeDaemonCore = async (
       }
     };
 
-    // Return a facade gateway whose `provide` method consults the
-    // current state of the remote-control on each call.  If the
-    // current connection is lost mid-flight (e.g., due to crossed
-    // hellos accept bias switching to a different TCP), subsequent
-    // `provide` calls will go through whatever connection the
-    // remote-control currently holds.
+    // TODO(option-a-simplification): After the dispose-callback change,
+    // the peer formula is destroyed on any connection loss, so this
+    // formula instance never sees a post-connect re-dial.  The
+    // `currentGatewayP` one-shot promise is still used for the initial
+    // connection (including the retention-set follower below) but the
+    // re-dial path in `ResilientPeerGateway.provide` is now unreachable.
+    // A follow-up can drop the ResilientPeerGateway wrapper and inline
+    // the single `dialAttempt()` call directly.
     const currentGatewayP = resilientDial();
 
     // Follow retention set changes in the background once connected.
@@ -5326,6 +5350,15 @@ const makeDaemonCore = async (
           // Try with the current gateway; on failure, re-dial and try
           // once more.  This handles the case where the initial dial
           // succeeded but the connection was later abandoned.
+          //
+          // TODO(option-a-simplification): The isAbandonError catch
+          // below is now unreachable for post-connect errors.  After
+          // the dispose-callback change (Option A), any connection loss
+          // cancels the peer formula's context, so this formula instance
+          // is torn down before `provide` could be called after a loss.
+          // The catch arm was the "retry within the existing formula
+          // instance" path; it is superseded by reincarnation.  A
+          // follow-up can remove the catch arm entirely.
           try {
             const gateway = await currentGatewayP;
             return await E(gateway).provide(requestedId);

--- a/packages/daemon/src/networks/iroh-heartbeat.js
+++ b/packages/daemon/src/networks/iroh-heartbeat.js
@@ -39,6 +39,26 @@ export const KEEPALIVE_TIMEOUT_MS = 2 * HEARTBEAT_INTERVAL_MS;
 harden(KEEPALIVE_TIMEOUT_MS);
 
 /**
+ * Render an unknown thrown value (which may not be an `Error`) for a log
+ * message without itself throwing. Mirrors the structure of `String(value)`
+ * but reaches for `.message` when one is present, so an `Error` renders as
+ * its message rather than `"Error"`.
+ *
+ * @param {unknown} value
+ * @returns {string}
+ */
+const renderThrown = value => {
+  if (value instanceof Error) {
+    return value.message;
+  }
+  try {
+    return String(value);
+  } catch {
+    return '<unrenderable thrown value>';
+  }
+};
+
+/**
  * Keep an iroh QUIC connection alive with a datagram heartbeat and detect a
  * dead peer with a keep-alive watchdog. The caller owns teardown: `onTimeout`
  * fires at most once, after the watchdog has been disarmed, and is expected to
@@ -50,6 +70,8 @@ harden(KEEPALIVE_TIMEOUT_MS);
  * @param {object} [options]
  * @param {number} [options.intervalMs] - Heartbeat send period.
  * @param {number} [options.timeoutMs] - Keep-alive window before `onTimeout`.
+ *   Defaults to twice the effective `intervalMs`, so the "tolerate a single
+ *   dropped beat" invariant holds for callers that only override `intervalMs`.
  * @param {() => void} [options.onTimeout] - Invoked once when the peer misses
  *   the keep-alive window.
  * @param {(message: string) => void} [options.log] - Diagnostic sink; silent
@@ -61,7 +83,7 @@ export const makeIrohHeartbeat = (
   connection,
   {
     intervalMs = HEARTBEAT_INTERVAL_MS,
-    timeoutMs = KEEPALIVE_TIMEOUT_MS,
+    timeoutMs = 2 * intervalMs,
     onTimeout = () => {},
     log = () => {},
   } = {},
@@ -121,9 +143,10 @@ export const makeIrohHeartbeat = (
     } catch (error) {
       // A full send buffer or transient datagram error is not fatal: the next
       // beat retries and the peer's watchdog tolerates a single miss.
-      log(
-        `iroh heartbeat send failed: ${/** @type {Error} */ (error).message}`,
-      );
+      // Stringify defensively: `sendDatagram` is native code and may throw a
+      // non-Error value, in which case reaching for `.message` would itself
+      // throw and crash the heartbeat loop.
+      log(`iroh heartbeat send failed: ${renderThrown(error)}`);
     }
   };
 

--- a/packages/daemon/test/iroh-heartbeat.test.js
+++ b/packages/daemon/test/iroh-heartbeat.test.js
@@ -193,3 +193,60 @@ test('a failing sendDatagram is caught and reported', async t => {
     'send failures surface through the log sink',
   );
 });
+
+test('a non-Error thrown by sendDatagram is rendered without re-throwing', async t => {
+  // Native bindings can throw arbitrary values, not just `Error` instances.
+  // Reaching for `.message` on a `null` throw would itself throw and crash
+  // the heartbeat loop; the loop must survive that case.
+  t.timeout(5000);
+  const logs = [];
+  let beats = 0;
+  const connection = {
+    sendDatagram() {
+      beats += 1;
+      // eslint-disable-next-line no-throw-literal
+      throw null;
+    },
+    readDatagram() {
+      return new Promise(() => {});
+    },
+  };
+  const heartbeat = makeIrohHeartbeat(connection, {
+    intervalMs: 20,
+    timeoutMs: 10_000,
+    log: message => logs.push(message),
+  });
+  t.teardown(() => heartbeat.stop());
+
+  await delay(80);
+  t.true(beats >= 2, 'the heartbeat loop survives a non-Error throw');
+  t.true(
+    logs.some(message => message.includes('heartbeat send failed')),
+    'a non-Error throw still surfaces through the log sink',
+  );
+});
+
+test('timeoutMs defaults to twice the chosen intervalMs', async t => {
+  // The documented "tolerate a single dropped beat" invariant must hold for
+  // callers that only override `intervalMs`. A 30ms interval should yield a
+  // ~60ms watchdog window, so silence past ~80ms must fire the timeout.
+  t.timeout(5000);
+  const connection = makeFakeConnection();
+  let timeouts = 0;
+  const heartbeat = makeIrohHeartbeat(connection, {
+    intervalMs: 30,
+    onTimeout: () => {
+      timeouts += 1;
+    },
+  });
+  t.teardown(() => heartbeat.stop());
+
+  // Arm the watchdog with one inbound beat, then let the peer fall silent.
+  connection.push(new Uint8Array([1]));
+  await delay(140);
+  t.is(
+    timeouts,
+    1,
+    'derived timeoutMs (2 * intervalMs) fires after the silence window',
+  );
+});

--- a/packages/daemon/test/peer-formula-revocation.test.js
+++ b/packages/daemon/test/peer-formula-revocation.test.js
@@ -1,0 +1,251 @@
+// @ts-nocheck
+/* global setTimeout */
+// Tests for Option A: peer formula context cancellation on connection loss.
+//
+// Option A changes the dispose callback in `makePeer` from:
+//   dropLiveValue(context.id)
+// to:
+//   context.cancel(new Error('peer connection lost'))
+//
+// This ensures that any connection loss cascades via `thisDiesIfThatDies`
+// to all dependent remote presences, revoking them.  The next use of any
+// remote presence reincarnates the peer formula and re-dials from scratch.
+//
+// These tests verify the lifecycle contract at the level of the context and
+// remote-control primitives, without spinning up a full daemon.
+
+import test from '@endo/ses-ava/prepare-endo.js';
+
+import harden from '@endo/harden';
+import { makePromiseKit as _makePromiseKit } from '@endo/promise-kit';
+
+import { makeContextMaker } from '../src/context.js';
+import { makeRemoteControlProvider } from '../src/remote-control.js';
+
+/** @import { PromiseKit } from '@endo/promise-kit' */
+
+/** @type {<T = never>() => PromiseKit<T>} */
+const makePromiseKit = _makePromiseKit;
+
+/** @typedef {import('../src/types.js').FormulaIdentifier} FormulaIdentifier */
+
+const id = /** @param {string} s @returns {FormulaIdentifier} */ s =>
+  /** @type {FormulaIdentifier} */ (s);
+
+/**
+ * Build a minimal context maker wired up to the real context implementation.
+ * Returns a `makeContext` factory and the shared `controllerForId` map.
+ */
+const setupContextMaker = () => {
+  /** @type {Map<FormulaIdentifier, { context: any }>} */
+  const controllerForId = new Map();
+  const formulaTypes = new Map();
+
+  const makeContext = makeContextMaker({
+    controllerForId,
+    provideController: formulaId => {
+      let controller = controllerForId.get(formulaId);
+      if (!controller) {
+        const ctx = makeContext(formulaId);
+        controller = { context: ctx };
+        controllerForId.set(formulaId, controller);
+      }
+      return controller;
+    },
+    getFormulaType: formulaId => formulaTypes.get(formulaId),
+  });
+
+  /**
+   * @param {FormulaIdentifier} formulaId
+   * @param {string} [type]
+   */
+  const createContext = (formulaId, type = 'test') => {
+    formulaTypes.set(formulaId, type);
+    const ctx = makeContext(formulaId);
+    controllerForId.set(formulaId, { context: ctx });
+    return ctx;
+  };
+
+  return { createContext, controllerForId };
+};
+
+// ---------------------------------------------------------------------------
+// Test 1: connection dispose triggers context.cancel on the peer formula.
+//
+// Simulates the Option A dispose callback shape:
+//   () => { context.cancel(new Error('peer connection lost')); }
+// The peer formula's `cancelled` promise must reject with that error.
+// ---------------------------------------------------------------------------
+test('connection dispose cancels the peer formula context', async t => {
+  t.timeout(5000);
+  const { createContext } = setupContextMaker();
+  const peerFormulaId = id('peer:local');
+  const peerContext = createContext(peerFormulaId, 'peer');
+
+  // Simulate the Option A dispose callback.
+  const disposeCallback = () => {
+    peerContext.cancel(new Error('peer connection lost'));
+  };
+
+  // Verify the context is alive before dispose.
+  let isCancelled = false;
+  peerContext.cancelled.catch(() => {
+    isCancelled = true;
+  });
+
+  // Fire the dispose callback (mimicking what remote-control does on loss).
+  disposeCallback();
+
+  // Wait for microtask queue to drain.
+  await Promise.resolve();
+
+  t.true(isCancelled, 'peer formula context must be cancelled after dispose');
+  await t.throwsAsync(() => peerContext.cancelled, {
+    message: 'peer connection lost',
+  });
+});
+
+// ---------------------------------------------------------------------------
+// Test 2: cancelling the peer formula cascades via thisDiesIfThatDies
+// to all dependent remote presences.
+//
+// `evaluateFormulaForId` registers remote formulas with:
+//   context.thisDiesIfThatDies(peerId)
+// so when the peer formula is cancelled, every remote presence is revoked.
+// ---------------------------------------------------------------------------
+test('peer formula cancellation cascades to dependent remote presences', async t => {
+  t.timeout(5000);
+  const { createContext } = setupContextMaker();
+
+  const peerFormulaId = id('peer:local');
+  const peerContext = createContext(peerFormulaId, 'peer');
+
+  // Two remote presences that depend on the peer formula.
+  const remotePresence1Id = id('rp1:local');
+  const remotePresence2Id = id('rp2:local');
+  const rp1Context = createContext(remotePresence1Id, 'remote-slot');
+  const rp2Context = createContext(remotePresence2Id, 'remote-slot');
+
+  // Wire dependencies: each remote presence dies if the peer formula dies.
+  // This matches evaluateFormulaForId: context.thisDiesIfThatDies(peerId).
+  rp1Context.thisDiesIfThatDies(peerFormulaId);
+  rp2Context.thisDiesIfThatDies(peerFormulaId);
+
+  // Option A dispose callback: cancel the peer formula's context.
+  // The cancel call itself is synchronous in its initial effects;
+  // cascade via thatDiesIfThisDies is also synchronous.
+  peerContext.cancel(new Error('peer connection lost'));
+
+  // Both remote presences must be revoked with the same error.
+  // The cascade is synchronous in the context implementation (cancel
+  // iterates dependents and calls their cancel immediately), so the
+  // cancelled promises are already rejected at this point.
+  await t.throwsAsync(() => rp1Context.cancelled, {
+    message: 'peer connection lost',
+  });
+  await t.throwsAsync(() => rp2Context.cancelled, {
+    message: 'peer connection lost',
+  });
+});
+
+// ---------------------------------------------------------------------------
+// Test 3: prior behavior (dropLiveValue only) did NOT cancel remote presences.
+//
+// This is a regression-evidence test: it demonstrates the exact failure mode
+// that Option A fixes.  Without context.cancel, `thisDiesIfThatDies` never
+// fires, so remote presences remain live after connection loss.
+// ---------------------------------------------------------------------------
+test('regression evidence: dropLiveValue alone does not cancel remote presences', async t => {
+  t.timeout(5000);
+  const { createContext, controllerForId } = setupContextMaker();
+
+  const peerFormulaId = id('peer:regression');
+  // peerContext is intentionally not captured: the regression test only needs
+  // the formula to be registered in the context map so that thisDiesIfThatDies
+  // can find it.  The returned context is not exercised directly.
+  createContext(peerFormulaId, 'peer');
+
+  const remotePresenceId = id('rp:regression');
+  const rpContext = createContext(remotePresenceId, 'remote-slot');
+
+  rpContext.thisDiesIfThatDies(peerFormulaId);
+
+  // Old dispose callback: only dropLiveValue (remove from cache, no cancel).
+  const oldDisposeCallback = () => {
+    controllerForId.delete(peerFormulaId);
+  };
+
+  oldDisposeCallback();
+
+  // The remote presence context must NOT be cancelled by dropLiveValue alone.
+  // We race its `cancelled` promise against a short delay; it should stay
+  // pending.
+  const raceResult = await Promise.race([
+    rpContext.cancelled.then(() => 'resolved').catch(() => 'rejected'),
+    new Promise(resolve => setTimeout(resolve, 50)).then(() => 'pending'),
+  ]);
+  t.is(
+    raceResult,
+    'pending',
+    'remote presence must remain live after dropLiveValue alone (pre-Option-A behavior)',
+  );
+});
+
+// ---------------------------------------------------------------------------
+// Test 4: remote-control dispose callback integration.
+//
+// The dispose callback is called by the remote-control state machine after
+// the connection's `cancelled` promise rejects (via `.then(currentDispose)` in
+// remote-control.js).  This test wires together the remote-control and the
+// Option A dispose callback to verify end-to-end cascade.
+// ---------------------------------------------------------------------------
+test('remote-control dispose triggers peer formula cancellation and cascades', async t => {
+  t.timeout(5000);
+  const { createContext } = setupContextMaker();
+
+  const peerFormulaId = id('peer:rc-test');
+  const peerContext = createContext(peerFormulaId, 'peer');
+
+  const remotePresenceId = id('rp:rc-test');
+  const rpContext = createContext(remotePresenceId, 'remote-slot');
+  rpContext.thisDiesIfThatDies(peerFormulaId);
+
+  // Option A dispose callback (what makePeer installs via remoteControl.connect).
+  const disposeCallback = () => {
+    peerContext.cancel(new Error('peer connection lost'));
+  };
+
+  const provideRemoteControl = makeRemoteControlProvider('local-node');
+  const remoteControl = provideRemoteControl('remote-node');
+
+  const { promise: connCancelled, reject: cancelConn } = makePromiseKit();
+
+  // A minimal fake gateway stub (the returned value is not exercised in
+  // this test; we only care that the dispose callback fires on cancel).
+  const fakeGateway = harden({});
+
+  // Install the dispose callback via remoteControl.connect.
+  remoteControl.connect(
+    () => fakeGateway,
+    cancelConn,
+    connCancelled,
+    disposeCallback,
+  );
+
+  // Simulate connection loss by cancelling the connection.
+  cancelConn(new Error('QUIC connection closed'));
+
+  // The remote-control calls disposeCallback after connCancelled rejects.
+  // The dispose is chained via `.then(currentDispose)` in remote-control.js
+  // so it fires asynchronously in the next microtask turn.
+  // Awaiting a short Promise.resolve() chain is sufficient to drain it.
+  await Promise.resolve();
+  await Promise.resolve();
+
+  await t.throwsAsync(() => rpContext.cancelled, {
+    message: 'peer connection lost',
+  });
+  await t.throwsAsync(() => peerContext.cancelled, {
+    message: 'peer connection lost',
+  });
+});

